src/StatsdClient/MetricSerializer.cs

```diff
@@ -0,0 +1,245 @@
+        {
+            return Metric.GetCommand(metricType, _prefix, name, value, sampleRate, _constantTags, tags);
+        }
+
+        private static string EscapeContent(string content)
+        {
+            return content
+                .Replace("\r", string.Empty)
+                .Replace("\n", "\\n");
+        }
+
+        private static string ConcatTags(string[] constantTags, string[] tags)
+        {
+            // avoid dealing with null arrays
+            constantTags = constantTags ?? EmptyStringArray;
+            tags = tags ?? EmptyStringArray;
+
+            if (constantTags.Length == 0 && tags.Length == 0)
+            {
+                return string.Empty;
+            }
+
+            var allTags = constantTags.Concat(tags);
+            string concatenatedTags = string.Join(",", allTags);
+            return $"|#{concatenatedTags}";
+        }
+
+        private static string TruncateOverage(string str, int overage)
+        {
+            return str.Substring(0, str.Length - overage);
+        }
+
+        public abstract class Metric
+        {
+            private static readonly Dictionary<MetricType, string> _commandToUnit = new Dictionary<MetricType, string>
+                                                                {
+                                                                    { MetricType.Counting, "c" },
+                                                                    { MetricType.Timing, "ms" },
+                                                                    { MetricType.Gauge, "g" },
+                                                                    { MetricType.Histogram, "h" },
+                                                                    { MetricType.Distribution, "d" },
+                                                                    { MetricType.Meter, "m" },
+                                                                    { MetricType.Set, "s" },
+                                                                };
+
+            public static string GetCommand<T>(MetricType metricType, string prefix, string name, T value, double sampleRate, string[] constantTags, string[] tags)
+            {
+                string full_name = prefix + name;
+                string unit = _commandToUnit[metricType];
+                var allTags = ConcatTags(constantTags, tags);
+
+                return string.Format(
+                    CultureInfo.InvariantCulture,
+                    "{0}:{1}|{2}{3}{4}",
+                    full_name,
+                    value,
+                    unit,
+                    sampleRate == 1.0 ? string.Empty : string.Format(CultureInfo.InvariantCulture, "|@{0}", sampleRate),
+                    allTags);
+            }
+        }
+
+        public class Event
+        {
+            private const int MaxSize = 8 * 1024;
```

Comment:
Should we make MaxSize configurable? I think I mentioned elsewhere, this 8Kb limit might not work well with remote UDP as it will likely require fragmentation which might be a bad idea.

Reply:
This is a very good comment.
`MaxSize` could be `StatsdMaxUDPPacketSize` from https://github.com/DataDog/dogstatsd-csharp-client/blob/master/src/StatsdClient/StatsdConfig.cs#L83, but if the message size cannot be reduced, the code throws an exception which is not caught (I can add`try` / `catch`).

I am wondering if the logic of trying reducing the message size is really useful. 
Indeed https://github.com/DataDog/dogstatsd-csharp-client/blob/master/src/StatsdClient/Bufferize/StatsBufferize.cs#L67-L70 already handle too long message. Adding an exception handler as in Java client will let the user know when something is wrong.

I think this change should be done in a separate PR, but I would love to hear your opinion about this topic.